### ISSUE TITLE
Building OpenSSL 3.0.5 - c1 : fatal error C1083: Cannot open source file: '': No such file or directory

### ISSUE THREAD

**Issue by jmbords** [openssl/openssl#18823]
As the title suggest, when building OpenSSL 3.0.5 source on Windows, I encounter error:

```
""cl"" /EP /C  -I"." -I"include" -I"providers\common\include" -I"providers\implementations\include" /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MDd /W3 /wd4090 /nologo /Od -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"C:\\Program Files (x86)\\Common Files\\SSL\"" -D"ENGINESDIR=\"C:\\openssl-3.0.5-debug\\lib\\engines-3\"" -D"MODULESDIR=\"C:\\openssl-3.0.5-debug\\lib\\ossl-modules\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"DEBUG" -D"_DEBUG"   -DAES_ASM -DCMLL_ASM -DDES_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_USE_APPLINK -DPOLY1305_ASM -DRMD160_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM crypto\aes\aes-586.asm.S > crypto\aes\aes-586.asm.i
```

```
c1 : fatal error C1083: Cannot open source file: '': No such file or directory
aes-586.asm.S
NMAKE : fatal error U1077: '""cl""' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\BIN\nmake.exe"' : return code '0x2'
Stop.
```

Commands I used:
```
set PATH=C:\Program Files (x86)\nasm;C:\perl\bin;%PATH%
perl Configure VC-WIN32 --prefix=C:/openssl-3.0.5-debug --debug
nmake
```

Note that because of the limitations of our project, I can only use Microsoft Visual Studio 10.0 _as of now_.
Is this a known issue?
Is this an issue on the compiler version or other configurations?

**Comment by jmbords** [comment]
I tried using Microsoft Visual Studio 19 and I still encountered the build error.
For context, I used the 3.0.5 package in the Downloads section [openssl.org](https://www.openssl.org/source/)
```
14721 | 2022-Jul-05 10:09:39 | openssl-3.0.5.tar.gz (SHA256) (PGP sign) (SHA1)
```

**Comment by jmbords** [comment]
Another note: I haven't encountered this issue on release 3.0.3.

**Comment by jmbords** [comment]
Here is another logs from nmake -d
```
Tue, Jul 05 2022 10:57:04               crypto\perlasm\x86asm.pl

crypto\aes\aes-586.asm target does not exist

cmd /C "set "ASM="nasm"" & "C:\oracle\ora1930\perl\bin\perl.exe" "crypto\aes\asm\aes-586.pl" "win32n" -I"." -I"include" -I"providers\common\include" -I"providers\implementations\include" /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MD /W3 /wd4090 /nologo /O2 -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"C:\\Program Files (x86)\\Common Files\\SSL\"" -D"ENGINESDIR=\"C:\\openssl-3.0.3-release\\lib\\engines-3\"" -D"MODULESDIR=\"C:\\openssl-3.0.3-release\\lib\\ossl-modules\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"NDEBUG"  -DAES_ASM -DCMLL_ASM -DDES_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_USE_APPLINK -DPOLY1305_ASM -DRMD160_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM  crypto\aes\aes-586.asm.S"

""cl"" /EP /C  -I"." -I"include" -I"providers\common\include" -I"providers\implementations\include" /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MD /W3 /wd4090 /nologo /O2 -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"C:\\Program Files (x86)\\Common Files\\SSL\"" -D"ENGINESDIR=\"C:\\openssl-3.0.3-release\\lib\\engines-3\"" -D"MODULESDIR=\"C:\\openssl-3.0.3-release\\lib\\ossl-modules\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"NDEBUG"   -DAES_ASM -DCMLL_ASM -DDES_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_USE_APPLINK -DPOLY1305_ASM -DRMD160_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM crypto\aes\aes-586.asm.S > crypto\aes\aes-586.asm.i

cl : Command line warning D9030 : '/EP' is incompatible with multiprocessing; ignoring /MP switch

c1: fatal error C1083: Cannot open source file: '': No such file or directory
aes-586.asm.S
NMAKE : fatal error U1077: '""cl""' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX86\x86\nmake.exe"' : return code '0x2'
Stop.
```

My colleague also noticed that aes-586.asm file doesn't exist on 3.0.5.
Only **aes-586.asm.S** and **aes-586.asm.i** exist.

**Comment by jmbords** [comment]
Based on the error logs and produced Makefile, my colleague removed the quotations on the CC value from the generated **Makefile**

Original Value:
```
CC="cl"
```

Edited Value:
```
CC=cl
```

After editing the CC value, the build finally succeeded.
As I understand, editing the Makefile is not recommended.
Any advice if we can use this generated build with this workaround?

**Comment by levitte** [comment]
Do you have an environment variable called CC?  What's its value in that case?  This command might tell:

```
echo %CC%
```

**Comment by jmbords** [comment]
CC variable is not set on our environment. If we use command
```
echo %CC%
```
the output only shows "ECHO is OFF".
CC also doesn't show up on "printenv" command.

**Comment by levitte** [comment]
Ah, I think I've found the culprit...

This:

https://github.com/openssl/openssl/blob/55461bf22a57a35d299f052323dfb659620a1dcd/Configurations/10-main.conf#L1312

... interacts with this:

https://github.com/openssl/openssl/blob/55461bf22a57a35d299f052323dfb659620a1dcd/Configurations/windows-makefile.tmpl#L254

The result of anything using `$(CPP)` is one set of quotes too many...  and .S files are typically treated using [ta-daaaaaaa] `$(CPP)`

Fix coming up.

**Comment by levitte** [comment]
It's curious, though, that this issue hasn't come up before.  This extra quotes thing has existed since december 2020.
I tested just now, and found that a fairly modern (Win10) cmd.exe tolerates that with zero issues, for example...  so I expect that this is version dependent.

Anyway, the extra quotes shouldn't be there, I think everyone will agree with that.

**Comment by levitte** [comment]
Fix in #18861, please try it out

**Comment by jmbords** [comment]
The changes worked now. I was able to nmake and nmake install it now.
Thanks for the quick fix.
